src/context/component-meta.js

```diff
@@ -0,0 +1,86 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {devAssert} from '../log';
+
+const EMPTY_DEPS = [];
+
+/**
+ * The following arguments are expected:
+ * - a required component function.
+ * - an optional array of `ContextProp` dependencies.
+ * - an optional string key for a global component identifier.
+ *
+ * Returns the provided function with metadata attached to it.
+ *
+ * @param {...?} args
+ * @return {function(...?):?}
+ */
+export function withMetaData(...args) {
```

Comment:
In a nutshell, what I really wanted is:

```
withMetaData(
  [Dep1, Dep2],
  (..., dep1, dep2) => {
  }
)
```

I.e. so that deps are right above the actual function signature so that it's much easier to track them next to each other. If it's on the bottom - a dev would have to scroll back up and down to see them.

The `key`, otoh, would be very rarely used, per my expectation.

Reply:
similar question then. can't we support it still with a more standard signature? e.g.:
`function withMetaData(deps: !Array<!ContextProp>|undefined, func: !Function, opt_key: string)`

The current implementation allows you to not pass deps if there aren't any, but I'd also expect if you didn't have deps you wouldn't be using `withMetaData` (since you stated folks will only need to set `key` rarely)

Reply:
Yeah. Thought about this some more and I agree. Done.